internal/app/readycheck.go

```diff
@@ -1,13 +1,152 @@
 
-}
+	clientv3 "go.etcd.io/etcd/client/v3"
+	"go.uber.org/zap"
+)
+
+const (
+	// ReadyServerPort is the port number for the server that serves the readiness probe
+	ReadyServerPort       = int64(9095)
+	etcdConnectionTimeout = 5 * time.Second
+	etcdGetTimeout        = 5 * time.Second
+	etcdQueryInterval     = 2 * time.Second
+	etcdEndpointAddress   = ":2379"
+)
 
-// SetupReadinessProbe sets up the readiness probe for this application.
+// SetupReadinessProbe sets up the readiness probe for this application. It is a blocking function and therefore
+// the consumer should always call this within a go-routine unless the caller itself wants to block on this which is unlikely.
 func (a *Application) SetupReadinessProbe() {
-	//the caller should call this in a go-routine as this method will setup a HTTP router and call ListenAndServe which blocks.
-	//If the http server errors out then you will have to panic and that should cause the container to exit and then be restarted by kubelet.
+	// Start go routine to regularly query etcd to update its readiness status.
+	go a.queryAndUpdateEtcdReadiness()
+	// If the http server errors out then you will have to panic and that should cause the container to exit and then be restarted by kubelet.
+	if a.isTLSEnabled() {
+		http.Handle("/readyz", http.HandlerFunc(a.readinessHandler))
+		err := http.ListenAndServeTLS(fmt.Sprintf(":%d", ReadyServerPort), a.cfg.ClientTLSInfo.CertFile, a.cfg.ClientTLSInfo.KeyFile, nil)
+		if err != nil {
+			a.logger.Fatal("failed to start TLS readiness endpoint", zap.Error(err))
+		}
+	} else {
+		http.Handle("/readyz", http.HandlerFunc(a.readinessHandler))
+		err := http.ListenAndServe(fmt.Sprintf(":%d", ReadyServerPort), nil)
+		if err != nil {
+			a.logger.Fatal("failed to start readiness endpoint", zap.Error(err))
+		}
+	}
+}
+
+// queryAndUpdateEtcdReadiness periodically queries the etcd DB to check its readiness and updates the status
+// of the query into the etcdStatus struct. It stops querying when the application context is cancelled.
+func (a *Application) queryAndUpdateEtcdReadiness() {
+	// Create a ticker to periodically query etcd readiness
+	ticker := time.NewTicker(etcdQueryInterval)
+	defer ticker.Stop()
+
+	for {
+		// Query etcd readiness and update the status
+		a.etcdReady = a.isEtcdReady()
+
+		select {
+		// Stop querying and return when the context is cancelled
+		case <-a.ctx.Done():
+			a.logger.Error("stopped periodic DB query: context cancelled", zap.Error(a.ctx.Err()))
+			return
+		// Wait for the next tick before querying again
+		case <-ticker.C:
+		}
+	}
+}
+
+// isEtcdReady checks if ETCD is ready by making a `GET` call (with a timeout).
+// if there is an error then it returns false else it returns true.
+func (a *Application) isEtcdReady() bool {
+	etcdConnCtx, cancelFunc := context.WithTimeout(a.ctx, etcdGetTimeout)
+	defer cancelFunc()
+	_, err := a.etcdClient.Get(etcdConnCtx, "foo", clientv3.WithSerializable())
```

Comment:
I think having `clientv3.WithSerializable()` is fair. Readiness should be specific to the pod itself and not the cluster as a whole.
Regarding your fear about one pod being marked `Ready` when quorum is lost (2/3 pods being not ready), it should be fine if 1/3 pod is marked ready. Etcd will anyway reject all write/read calls because quorum is missing

Reply:
As decided, we will revisit the readinessProbe after 1 release of etcd-wrapper but for now please rollback this readinessProbe with `clientv3.WithSerializable()`
https://github.com/gardener/etcd-wrapper/issues/7